src/Pulsar.Client/Auth/TokenClient.fs

```diff
@@ -0,0 +1,82 @@
+﻿module Pulsar.Client.Auth.Oauth2Token
+
+open System
+open System.Collections.Generic
+open System.Net
+open System.Net.Http
+open System.Net.Http.Headers
+open System.Text.Json.Serialization
+open System.Text.Json
+
+type TokenError =
+    {    
+        [<JsonPropertyName("error")>]
+        Error:string 
+        
+        [<JsonPropertyName("error_description")>]
+        ErrorDescription:string 
+        
+        [<JsonPropertyName("error_uri")>]
+        ErrorUri:string 
+    }
+
+type TokenResult =
+    {    
+        [<JsonPropertyName("access_token")>]
+        AccessToken:string
+        
+        [<JsonPropertyName("id_token")>]
+        IdToken:string
+        
+        [<JsonPropertyName("refresh_token")>]
+        RefreshToken:string
+        
+        [<JsonPropertyName("expires_in")>]
+        ExpiresIn:int 
+    }
+
+type TokenExchangeResult =
+    | Result of TokenResult * DateTime
+    | OauthError of TokenError
+    | OtherError of string
+    
+
+let Exchange (uri:Uri,clientId:string, clientSecret:string, audience:Uri) : TokenExchangeResult =
+     async{
```

Comment:
TaskBuilder here as well. Also please change function name to lowercase and make parameters to carried form